### PR TITLE
Concept diagnostics

### PR DESCRIPTION
**Pull request by gowthamrao**
Optimize concept set diagnostics

This PR replaces the older concept set diagnostics (that had a few bugs and surprises) to an optimized version

- in this PR, the movement of files from db to local csv is minimized
- bulk orphan logic that performs it for all concept set
- addition of exclusion module

### CODE REVIEW

File: R/CohortComparisonDiagnostics.R [hunk 1]
```diff
@@ -67,18 +67,15 @@ computeCohortOverlap <- function(connectionDetails = NULL,
     return(tidyr::tibble())
   }
   
-  sql <- SqlRender::loadRenderTranslateSql(
-    "CohortOverlap.sql",
-    packageName = utils::packageName(),
-    dbms = connection@dbms,
+  sql <- SqlRender::readSql(system.file("sql/sql_server/CohortOverlap.sql",
```
Comment (azimov) [inline R/CohortComparisonDiagnostics.R:70]: instead of slashes use separate strings e.g.

```
system.file("sql", "sql_server", CohortOverlap.sql", package = utils::packageName())
```

File: R/CohortLevelDiagnostics.R [hunk 2]
```diff
@@ -42,20 +42,18 @@ getCohortCounts <- function(connectionDetails = NULL,
...
-  sql <-
-    SqlRender::loadRenderTranslateSql(
-      sqlFilename = "CohortCounts.sql",
-      packageName = utils::packageName(),
-      dbms = connection@dbms,
-      cohort_database_schema = cohortDatabaseSchema,
-      cohort_table = cohortTable,
-      cohort_ids = cohortIds
-    )
+  sql <- SqlRender::readSql(system.file("sql/sql_server/CohortCounts.sql",
```
Comment (azimov) [inline R/CohortLevelDiagnostics.R:45]: again, remove use of slashes in file paths

File: R/IncidenceRates.R [hunk 3]
```diff
@@ -56,15 +56,13 @@ getIncidenceRate <- function(connectionDetails = NULL,
...
   
   ParallelLogger::logInfo("Calculating incidence rate per year by age and gender")
-  sql <-
-    SqlRender::loadRenderTranslateSql(
-      sqlFilename = "GetCalendarYearRange.sql",
-      packageName = utils::packageName(),
-      dbms = connection@dbms,
-      cdm_database_schema = cdmDatabaseSchema
-    )
+  sql <- SqlRender::readSql(system.file("sql/sql_server/GetCalendarYearRange.sql",
```
Comment (azimov) [inline R/IncidenceRates.R:59]: Code changes in this file are unrelated to the PR. Doing this increases the chances of a bug entering in the codebase

File: R/ResultsDataModel.R [hunk 4]
```diff
@@ -255,12 +255,16 @@ createResultsDataModel <- function(connection = NULL,
...
     stop("Invalid schema for sqlite, use schema = 'main'")
   }
-
-  sql <- SqlRender::loadRenderTranslateSql(sqlFilename = "CreateResultsDataModel.sql",
-                                           packageName = utils::packageName(),
-                                           dbms = connection@dbms,
-                                           results_schema = schema)
-  DatabaseConnector::executeSql(connection, sql)
+  
+  sql <- SqlRender::readSql(system.file("sql/sql_server/CreateResultsDataModel.sql",
```
Comment (azimov) [inline R/ResultsDataModel.R:259]: Again, this refactoring is unnecessary and will probably not be tested during development. Its not a major change but why add it here?

File: R/ResultsDataModel.R [hunk 5]
```diff
@@ -531,6 +535,16 @@ deleteAllRecordsForDatabaseId <- function(connection,
                                           schema,
                                           tableName,
                                           databaseId) {
+  tableExists <-
```
Comment (azimov) [inline R/ResultsDataModel.R:538]: There are no guarantees this will work and it doesn't seem related to the PR

File: inst/sql/sql_server/CreateResultsDataModel.sql [hunk 6]
```diff
@@ -31,95 +37,42 @@
...
-
-IF OBJECT_ID('@results_schema.@time_distribution', 'U') IS NOT NULL
-    DROP TABLE @results_schema.@time_distribution;
-
-IF OBJECT_ID('@results_schema.@visit_context', 'U') IS NOT NULL
-    DROP TABLE @results_schema.@visit_context;
-
-IF OBJECT_ID('@results_schema.@vocabulary', 'U') IS NOT NULL
-    DROP TABLE @results_schema.@vocabulary;
+DROP TABLE IF EXISTS @results_schema.@cohort;
```
Comment (azimov) [inline inst/sql/sql_server/CreateResultsDataModel.sql:40]: This is postgres specific syntax and will break most databasae platforms. Its also not related to the scope of the PR.

File: inst/sql/sql_server/OptimizeConceptSet.sql [hunk 7]
```diff
@@ -0,0 +1,195 @@
+DROP TABLE IF EXISTS #o_not_excluded;
```
Comment (azimov) [inline inst/sql/sql_server/OptimizeConceptSet.sql:1]: Again, drop table if exists is postgres specific syntax. This will break most for most platforms

File: R/IncidenceRates.R (hunk 3 above)
Comment (gowthamrao) [reply R/IncidenceRates.R:59]: It can - but we can test it. Reason i am making the change here is to ensure we have the same pattern everywhere. Consistent pattern will help us debug. This pattern is being used everywhere

File: R/ResultsDataModel.R (hunk 4 above)
Comment (gowthamrao) [reply R/ResultsDataModel.R:259]: So the reason is 

`SqlRender::loadRenderTranslateSql` has the argument dbms. The default value is "sql_server". But notice that in prior code there is dbms = connection@dbms . So when connection was to sqlite it was looking for code in the path sqlite instead of sql_server. 

I found it confusing as it is an implicit and not explicit. This was implicit, because we always use the sql_server as the core folder for SQL. 

This code change is to make sure we are explicit and consistent in our pattern everywhere in the package (i.e. not just in the code for this PR).

File: R/ResultsDataModel.R (hunk 5 above)
Comment (gowthamrao) [reply R/ResultsDataModel.R:538]: I can take that off - thank you.

ofcourse if there is no table - the sql will fail. we just have to make sure that deleteAllRecordsForDatabaseId is not used without first knowing there is such a table.

File: inst/sql/sql_server/CreateResultsDataModel.sql (hunk 6 above)
Comment (gowthamrao) [reply inst/sql/sql_server/CreateResultsDataModel.sql:41]: Martijn dropped this new knowledge for me - its supported sqlRender 1.9+

image

File: R/ResultsDataModel.R (hunk 4 above)
Comment (azimov) [reply R/ResultsDataModel.R]: > SqlRender::loadRenderTranslateSql has the argument dbms. The default value is "sql_server". But notice that in prior code there is dbms = connection@dbms . So when connection was to sqlite it was looking for code in the path sqlite instead of sql_server.

This is confusing to me; the behaviour of `loadRenderTranslateSql` should look for the `sqlite` folder first then the `sql_server` folder. The reason for this is that there may be code that is specific to `sqlite` for which the translation doesn't work. In which case, you would add folders and sql files specific for that db platform. See the[ common data model ](https://github.com/OHDSI/CommonDataModel/tree/master/inst/ddl/5.4)for an example of this.

File: R/CohortComparisonDiagnostics.R (hunk 1 above)
Comment (azimov) [reply R/CohortComparisonDiagnostics.R]: Use of sql_server folder directly is not advised, the way `loadRenderTranslateSql` worked before is correct.

File: R/ResultsDataModel.R (hunk 4 above)
Comment (gowthamrao) [reply R/ResultsDataModel.R]: But i dont get that same error in develop branch - only in the code for this PR. i dont understand why

This works fine
https://github.com/OHDSI/CohortDiagnostics/blob/develop/extras/tests/TestCodeEunomiaPackageMode.R

Comment (azimov) [reply R/ResultsDataModel.R]: The problem you're having is an interaction between devtools and SqlRender's file loader.

This is caused by the way that devtools test works; the normal mode of testing calls an actual package installed in the system but devtools uses a `load_all()` function. This can be solved by using rstudio for tests or creating a shortcut to the `<PackageRoot>/inst/sql` folder: `<PackageRoot>/sql`. This is resolved in linux/mac systems in the `test/testthat/helpers.R` but the solution used there (creating an sql symbolic link/shortcut when the tests run) does not work on windows because of permissions issues on many user's platforms.

File: R/ConceptSetDiagnostics.R [hunk 8]
```diff
@@ -0,0 +1,2295 @@
...
+                    .data$cohortId,
+                    .data$conceptSetId) %>%
+    dplyr::arrange(.data$uniqueConceptSetId,
+                   .data$cohortId,
+                   .data$conceptSetId)
+  return(conceptSets)
+}
+
+
+mergeTempTables <-
```
Comment (azimov) [inline R/ConceptSetDiagnostics.R:721]: Looks like the tests are failing with this function

File: R/ConceptSetDiagnostics.R [hunk 9]
```diff
@@ -0,0 +1,2295 @@
...
+  function(connection,
+           tableName,
+           tempTables,
+           tempEmulationSchema) {
+    valueString <-
+      paste(tempTables, collapse = "\n\n  UNION ALL\n\n  SELECT *\n  FROM ")
+    sql <-
+      sprintf("SELECT *\nINTO %s\nFROM (\n  SELECT *\n  FROM %s\n) tmp;",
```
Comment (azimov) [inline R/ConceptSetDiagnostics.R:729]: what's the reason for using sprintf instead of `renderTranslateExecute`?

File: R/ConceptSetDiagnostics.R [hunk 10]
```diff
@@ -0,0 +1,2295 @@
...
+                           tempEmulationSchema = tempEmulationSchema)
+    DatabaseConnector::executeSql(connection,
+                                  sql,
+                                  progressBar = FALSE,
+                                  reportOverallTime = FALSE)
+    
+    # Drop temp tables:
+    for (tempTable in tempTables) {
+      sql <-
+        sprintf("TRUNCATE TABLE %s;\nDROP TABLE %s;", tempTable, tempTable)
```
Comment (azimov) [inline R/ConceptSetDiagnostics.R:744]: use `renderTranslateExecuteSql`

File: R/ConceptSetDiagnostics.R (hunk 8 above)
Comment (gowthamrao) [reply R/ConceptSetDiagnostics.R:721]: @azimov  the mergeTempTables function was never changed 

see

https://github.com/OHDSI/CohortDiagnostics/blob/main/R/ConceptSets.R#L171-L204

File: R/ConceptSetDiagnostics.R (hunk 9 above)
Comment (gowthamrao) [reply R/ConceptSetDiagnostics.R:729]: Just how it was written the first time this function was written two years ago. it is the version in main branch.

File: R/ConceptSetDiagnostics.R (hunk 10 above)
Comment (gowthamrao) [reply R/ConceptSetDiagnostics.R:744]: Same comment - any change would be unrelated refactor
